File: HeartBeatSystemDemo-java/HeartBeatSystemDemo/src/thread/DragThread.java
package thread;

import iface.Drags;
import models.HeartBeatSystem;

import java.util.LinkedList;
import java.util.Timer;

public class DragThread implements Runnable{
    private HeartBeatSystem hbs;
    private LinkedList<Drags> dragList;
    private Timer timer;
    private int counter;
    public DragThread(HeartBeatSystem hbs){
        this.hbs = hbs;
        timer = new Timer();
        updateDragList();
        counter = 0;
    }

    /**
     * 这个方法用于更新计药物列表
     * 每次task执行时都会被调用
     */
    private void updateDragList(){
        dragList = hbs.getDragList();
        /*
        if(dragList.isEmpty()){
            return;
        }
        for(Drags drag : dragList){
            if(drag.getDuration() > maxDuration){
                maxDuration = drag.getDuration();
            }
        }
        */
    }

    /**
     * 该线程直接从hbs对象中启动，一旦启动就不再关闭
     */
    @Override
    public void run(){
        TimerTask task = new TimerTask();
        timer.schedule(task, 0 , 1000);
    }
    private class TimerTask extends java.util.TimerTask{
        @Override
        public void run() {
            if(dragList.isEmpty()){
                counter = 0;
            }
            System.out.println(counter/1000 + " second");
            for(Drags d : dragList){
                if(d.getDuration() <= counter){
                    hbs.removeDrag(d);
                    updateDragList();
                }
                counter = counter + 1000;
            }
        }
    }
}

File: UnityProjet/scripts/HeartBeatSystem/thread/DragThread.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Threading;
using UnityEngine;

public class DragThread
{
    private HeartBeatSystem HBS;
    private LinkedList<IDrags> DragList;
    private int Counter;
    private Boolean IsPause;//Pause lock when game pause the thread will not work
    public DragThread(HeartBeatSystem h)
    {
        this.HBS = h;
        Counter = 0;
        IsPause = false;
    }
    public void Pause()
    {
        IsPause = true;
    }
    public void UnPause()
    {
        IsPause = false;
    }
    private void UpdateDragList()
    {
        this.DragList = HBS.GetDragList();
    }
    public void Run()
    {
        while (true)
        {
            if (IsPause)
            {
                continue;
            }
            if (DragList.Count == 0)
            {
                Counter = 0;
            }
            Debug.Log(Counter / 1000 + " second");
            foreach (IDrags d in DragList)
            {
                if (d.GetDuration() <= Counter)
                {
                    HBS.RemoveDrag(d);
                    UpdateDragList();
                }
                Counter = Counter + 1000;
            }
            Thread.Sleep(1000);
        }
    }
}
